src/irmin-pack/unix/errors.ml

```diff
@@ -14,42 +14,83 @@
+  | `Invalid_layout
+  | `Corrupted_legacy_file
+  | `Pending_flush
+  | `Rw_not_allowed
+  | `Migration_needed
+  | `Corrupted_control_file
+  | `Sys_error of string
+  | `V3_store_from_the_future
+  | `Gc_forbidden_during_batch
+  | `Unknown_major_pack_version of string
+  | `Only_minimal_indexing_strategy_allowed
+  | `Commit_key_is_indexed_and_dangling of string
+  | `Dangling_key of string
+  | `Gc_disallowed
+  | `Node_or_contents_key_is_indexed of string
+  | `Commit_parent_key_is_indexed of string ]
+[@@deriving irmin ~pp]
+(** [base_error] is the type of most errors that can occur in a [result], except
+    for errors that have associated exceptions (see below) and backend-specific
+    errors (see {!Io_errors}). *)
 
-module type S = sig
-  (** Error manager for [RO_not_allowed], [Pack_error], [Io.misc_error] and IO's
-      dedicated exceptions *)
+type read_only_error = [ `Ro_not_allowed ] [@@deriving irmin ~pp]
 
-  module Io : Io.S
+type closed_error = [ `Closed | `Read_on_closed | `Write_on_closed ]
+[@@deriving irmin ~pp]
 
-  type t = [ base_error | `Io_misc of Io.misc_error | `Ro_not_allowed ]
+type error = [ base_error | closed_error | read_only_error ]
 
-  val pp : Format.formatter -> [< t ] -> unit
-  val raise_error : [< t ] -> 'a
-  val catch : (unit -> 'a) -> ('a, [> t ]) result
-  val raise_if_error : ('a, [< t ]) result -> 'a
-end
+exception Pack_error of base_error
+exception RO_not_allowed = Irmin_pack.RO_not_allowed
+exception Closed = Irmin.Closed
+
+(** Error manager *)
+module type S = sig
+  type t
 
-module Make (Io : Io.S) : S with module Io = Io = struct
-  module Io = Io
+  val pp : Format.formatter -> t -> unit
+  val raise_error : t -> 'a
+  val catch : (unit -> 'a) -> ('a, t) result
+  val raise_if_error : ('a, t) result -> 'a
+end
 
-  type t = [ base_error | `Io_misc of Io.misc_error | `Ro_not_allowed ]
-  type misc_error = Io.misc_error [@@deriving irmin ~pp]
+let pp ppf = function
+  | #read_only_error as e -> Format.fprintf ppf "%a" pp_read_only_error e
+  | #closed_error as e -> Format.fprintf ppf "%a" pp_closed_error e
+  | #base_error as e -> Format.fprintf ppf "%a" pp_base_error e
 
-  let pp ppf = function
-    | `Ro_not_allowed -> Format.fprintf ppf "`Ro_not_allowed"
-    | `Io_misc e -> Format.fprintf ppf "%a" pp_misc_error e
-    | #base_error as e -> Format.fprintf ppf "%a" pp_base_error e
+let raise_error = function
+  | #read_only_error -> raise Irmin_pack.RO_not_allowed
+  | #closed_error -> raise Irmin.Closed
+  | #base_error as e -> raise (Pack_error e)
 
-  let raise_error = function
-    | `Io_misc e -> Io.raise_misc_error e
-    | `Ro_not_allowed -> raise RO_not_allowed
-    | #base_error as e -> raise (Pack_error e)
+let catch f =
+  try f () with
+  | Pack_error e -> Error (e : base_error :> [> base_error ])
+  | RO_not_allowed -> Error `Ro_not_allowed
+  | Closed -> Error `Closed
+(* TODO: is this loss of infomration acceptable when going from exn to err? *)
```

Comment:
I had to add a `Closed` error for this exception catch. Not sure how I feel about it so flagged it.

Reply:
@Ngoguey42 what are your thoughts on this?

Reply:
Ah, I forgot to remove the read and write closed errors from `error_base`! Will do. Unless you're thinking differently now, I will keep double close in the base errors (as we discussed last week) since it does not fit with `Irmin.Closed`'s documentation explicitly excluding double close.

Reply:
Sure 👍

Reply:
Okay, just pushed up changes that removed those errors. We only have `Closed` now! :tada: